scripts/generate_datasource.go

```diff
@@ -0,0 +1,359 @@
+		r := &schema.Resource{
+			Schema: map[string]*schema.Schema{},
+		}
+		for n, p := range p.Properties {
+			se, err := getSchemaFromProperty(name+"."+n, p)
+			if err != nil {
+				return nil, err
+			}
+			if se == nil {
+				continue
+			}
+			r.Schema[n] = se
+		}
+		s.Elem = r
+	case "array":
+		s.Type = schema.TypeList
+		if len(p.Items.Schema.Properties) != 0 {
+			r := &schema.Resource{
+				Schema: map[string]*schema.Schema{},
+			}
+			for n, p := range p.Items.Schema.Properties {
+				se, err := getSchemaFromProperty(name+"."+n, p)
+				if err != nil {
+					return nil, err
+				}
+				if se == nil {
+					continue
+				}
+				r.Schema[n] = se
+			}
+			s.Elem = r
+		} else {
+			se, err := getSchemaFromProperty(name, *p.Items.Schema)
+			if err != nil {
+				return nil, err
+			}
+			se.Optional = false
+			se.Computed = false
+			s.Elem = se
+		}
+	default:
+		return nil, fmt.Errorf("unknwon type %q for %s", ty, name)
+	}
+
+	return s, nil
+}
+
+func (r *Resource) Render() (string, error) {
+	attrs := map[string]schema.ValueType{}
+	for name, schema := range r.schema {
+		if schema.Optional {
+			attrs[name] = schema.Type
+		}
+	}
+	attrs_ := []string{}
+	for a := range attrs {
+		attrs_ = append(attrs_, a)
+	}
+	sort.Strings(attrs_)
+
+	imports := map[string]struct{}{
+		"context": {},
+		"net/url": {},
+	}
+
+	var queryParams []string
+	for _, attr := range attrs_ {
+		switch ty := attrs[attr]; ty {
+		case schema.TypeBool:
+			imports["strconv"] = struct{}{}
+			queryParams = append(queryParams, fmt.Sprintf(`%s := d.Get(%q).(bool)`, attr, attr))
+			queryParams = append(queryParams, fmt.Sprintf(`if %s {`, attr))
+			queryParams = append(queryParams, fmt.Sprintf(`q.Add(%q, strconv.FormatBool(%s))`, attr, attr))
+			queryParams = append(queryParams, `}`)
+		case schema.TypeString:
+			queryParams = append(queryParams, fmt.Sprintf(`q.Add(%q, d.Get(%q).(string))`, attr, attr))
+		default:
+			panic(fmt.Sprintf("unknown type %q for %q", ty, attr))
+		}
+
+	}
+
+	imports_ := []string{}
+	for i := range imports {
+		imports_ = append(imports_, i)
+	}
+	sort.Strings(imports_)
+
+	t := template.Must(template.New("datasource").Parse(`
+// Code generated by scripts/generate_datasource.go. DO NOT EDIT.
+//go:{{.GenerateLine}}
+
+// This file was generated based on Boundary {{.Version}}
+
+package provider
+
+import (
+	{{range .Imports}}"{{.}}"
+	{{end}}
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+var dataSource{{.Name}}Schema = {{.Schema}}
+
+func dataSource{{.Name}}() *schema.Resource{
+	return &schema.Resource{
+		Description: {{.Description}},
+		Schema: dataSource{{.Name}}Schema,
+		ReadContext: dataSource{{.Name}}Read,
+	}
+}
+
+func dataSource{{.Name}}Read(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*metaData).client
+
+	req, err := client.NewRequest(ctx, "GET", "{{.Path}}", nil)
```

Comment:
This resolves my comment about using the API above.

Reply:
Indeed, as the OpenAPI document gives exactly what we need to send, it's easier to just use an `http.Client` rather than having to understand how the generator for the Go SDK work to predict the name of the types and methods to call.

I tried briefly to use the Go SDK but it makes the generator much messier than what we have here.